### ISSUE TITLE
LU factorization in the linalg extension

### ISSUE THREAD

**Issue by ogrisel** [data-apis/array-api#627]
It seems that many libraries that are candidates to implement the Array API namespace already implement the LU factorization (with variations in API and with the notable exception of numpy).

However LU is not part of the list of linear algebra operations of the current state of the SPEC:

- https://data-apis.org/array-api/latest/extensions/linear_algebra_functions.html#objects-in-api

Are there any plans to consider it for inclusion?

**Comment by rgommers** [comment]
Thanks for asking @ogrisel. I had a look through the initial issues which considered the various `linalg` APIs, and LU decomposition was not considered at all there. The main reason I think being that the overview started from what is present in NumPy, and then looked at matching APIs in other libraries.

I think it's an uphill battle for now. It would require adding it to `numpy.linalg`, moving it in other libraries with numpy-matching APIs (e.g., https://docs.cupy.dev/en/stable/reference/scipy_linalg.html#decompositions is in the wrong place), and then aligning on APIs also with PyTorch & co. Finally, there's `lu` but also `lu_solve` and `lu_factor` - would it be just one of those, or 2/3?

It seems to me that LU decomposition is important enough that it's worth working on. So we could figure out what the optimal API for it would be, and then adding it to `array-api-compat` so it can be used in scikit-learn and SciPy. That can be done on pretty short notice I think. From there to actually standardizing it would take quite a long time I suspect (but nothing is really blocked on not having that done).

**Comment by rgommers** [comment]
The signatures to consider:
- [SciPy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.lu_factor.html)/`cupyx.scipy`/`jax.scipy`: `lu(a, permute_l=False, overwrite_a=False, check_finite=True)`
- [PyTorch](https://pytorch.org/docs/stable/generated/torch.linalg.lu.html): `torch.linalg.lu(A, *, pivot=True, out=None)`

The `overwrite_a`, `check_finite` and `out` keywords should all be out of scope for the standard. 

The `permute_l`/`pivot` keywords do seem relevant to include. They control the return values in a different way. SciPy's `permute_l` returns 3 arrays if False, 2 arrays if True. That breaks a key design rule for the array API standard (no polymorphic APIs), so we can't do that. The PyTorch `pivot=True` behavior is okay, it always returns: _a named tuple `(P, L, U)`_, and leaves `P` as an empty array for the non-default `pivot=False`.

PyTorch defaults to partial pivoting, and the keyword allows no pivoting. An LU decomposition with full pivoting is also a thing mathematically, but it does not seem implemented. JAX also has [jax.lax.linalg.lu](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.linalg.lu.html), which only does partial pivoting.

So it seems like `lu(x, /) -> tuple(array, array, array):` which defaults to partial pivoting is the minimum needed, the question is whether the other pivoting mode(s) is/are needed.

**Comment by rgommers** [comment]
[dask.array.linalg.lu](https://docs.dask.org/en/stable/generated/dask.array.linalg.lu.html#dask-array-linalg-lu) has no keywords at all, and no info in the docstring about what is implemented. From the tests it's clear that it matches the SciPy default (`permute_l=False`).

**Comment by rgommers** [comment]
For PyTorch, the non-default flavor is only implemented on GPU:
```python
>>> A = torch.randn(3, 2)
... P, L, U = torch.linalg.lu(A)
>>> A = torch.randn(3, 2)
... P, L, U = torch.linalg.lu(A, pivot=False)
Traceback (most recent call last):
  Cell In[6], line 2
    P, L, U = torch.linalg.lu(A, pivot=False)
RuntimeError: linalg.lu_factor: LU without pivoting is not implemented on the CPU
```
Its docstring also notes: _The LU decomposition without pivoting [may not exist](https://en.wikipedia.org/wiki/LU_decomposition#Definitions) if any of the principal minors of A is singular._

tl;dr maybe the best way to go is to only implement partial pivoting?

**Comment by ogrisel** [comment]
Maybe we can start with a function with no argument that always returns PLU (that is only implement scipy's `permute_L=False` and torch's `pivot=True`) and it will be up to the consumer to compute. 

On the other hand, I think it would be good to have an option wot do the `PL` product automatically and avoid allocating P. Should array API expose two methods? `xp.linalg.lu` that outputs a 3-tuple `(P, L, U)` a second function `xp.linalg.permuted_lu` that precomputes the PL product and always outputs a 2-tuple `(P @ L, U)`?

> Its docstring also notes: The LU decomposition without pivoting [may not exist](https://en.wikipedia.org/wiki/LU_decomposition#Definitions) if any of the principal minors of A is singular.

Also note, from PyTorch's doc:

> The LU decomposition is almost never unique, as often there are different permutation matrices that can yield different LU decompositions. As such, different platforms, like SciPy, or inputs on different devices, may produce different valid decompositions.

Such a disclaimer should probably be mentioned in the Array API spec.

**Comment by ogrisel** [comment]
We do use the "permute_l=True" case in scikit-learn.

**Comment by ogrisel** [comment]
It would be easy to provide a fallback implementation that uses an extra temporary allocation + mm product for libraries that do not natively support scipy's permute_l=True.

But it's not clear if pytorch' pivot=False is equivalent to scipy permute_l=True or doing something different.

**Comment by rgommers** [comment]
> We do use the "permute_l=True" case in scikit-learn.

Oops yes, I didn't read well.

> But it's not clear if pytorch' pivot=False is equivalent to scipy permute_l=True or doing something different.

It looks like it's doing the exact same thing:
```python
>>> import torch
>>> import numpy as np
>>> from scipy import linalg

>>> t = torch.rand(20, 10, device='cuda')  # must be 'cuda', because `pivot=False` is not supported on CPU
>>> Pt, Lt, Ut = torch.linalg.lu(t.to('cuda'), pivot=False)
>>> Lt.shape
torch.Size([20, 10])
>>> Ut.shape
torch.Size([10, 10])
>>> Pt
tensor([], device='cuda:0')

>>> Ln, Un = linalg.lu(t.to('cpu').numpy(), permute_l=True)
>>> Ln.shape
(20, 10)
>>> Un.shape
(10, 10)
>>> np.testing.assert_allclose(Lt.to('cpu'), Ln)
...
AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0

Mismatched elements: 190 / 200 (95%)
Max absolute difference: 56.958973
Max relative difference: 181.88818

>>> (Lt @ Ut - t).max()
tensor(2.1011e-06, device='cuda:0')
>>> (Ln @ Un - t.to('cpu').numpy()).max()
1.1920929e-07
```
So both give valid decompositions with the same shapes for `L` and `U`, and `P` either empty or not returned; numerical values are very different but that's expected.

One other thing to point out, the non-pivoting case does result in a decomposition with larger errors for PyTorch:
```python
>>> Pt, Lt, Ut = torch.linalg.lu(t)
>>> (Pt @ Lt @ Ut - t).abs().max()
tensor(1.4901e-07, device='cuda:0')
>>> Pt, Lt, Ut = torch.linalg.lu(t, pivot=False)
>>> (Lt @ Ut - t).abs().max()
tensor(2.1011e-06, device='cuda:0')

>>> Pn, Ln, Un = linalg.lu(x)
>>> np.abs((Pn @ Ln @ Un - x)).max()
1.1920929e-07
>>> PLn, Un = linalg.lu(x, permute_l=True)
>>> np.abs((PLn @ Un - x)).max()
1.1920929e-07
```

SciPy's pivoting decomposition seems to be a lot faster for small arrays, and has similar performance for large arrays when they are square or wide, and is several times slower when they're tall:
```
>>> %timeit linalg.lu(x)  # shape (20, 10)
8.68 µs ± 38.8 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
>>> %timeit linalg.lu(x, permute_l=True)
21.7 µs ± 80.2 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

>>> %timeit linalg.lu(x)  # shape (400, 300)
951 µs ± 2.27 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
>>> %timeit linalg.lu(x, permute_l=True)
832 µs ± 7.62 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
>>> %timeit Pn @ Ln
190 µs ± 42.7 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

>>> x = np.random.randn(1000, 1000)
>>> %timeit linalg.lu(x)
8.49 ms ± 9.89 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
>>> %timeit linalg.lu(x, permute_l=True)
7.94 ms ± 19.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

>>> x = np.random.randn(1000, 50)  # tall array
>>> %timeit linalg.lu(x)
734 µs ± 2.21 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
>>> %timeit linalg.lu(x, permute_l=True)
250 µs ± 2.21 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
>>> x = np.random.randn(50, 1000)  # wide array
>>> %timeit linalg.lu(x)
220 µs ± 303 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
>>> %timeit linalg.lu(x, permute_l=True)
222 µs ± 4.48 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

Given that scikit-learn does use `permute_l=True` and I suspect the tall array case is quite common for ML applications, there does seem to be a case for providing both flavors.

**Comment by rgommers** [comment]
For completeness, some PyTorch GPU timings:
```python
>>> t = torch.rand(1000, 1000, device='cuda')  # square tensor
>>> %timeit torch.linalg.lu(t)
7.11 ms ± 382 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
>>> %timeit torch.linalg.lu(t, pivot=False)
1.28 ms ± 2.47 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

>>> t = torch.rand(1000, 50, device='cuda')  # tall tensor
>>> %timeit torch.linalg.lu(t)
2.15 ms ± 54.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
>>> %timeit torch.linalg.lu(t, pivot=False)
343 µs ± 6.58 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

>>> t = torch.rand(50, 1000, device='cuda')  # wide tensor
>>> %timeit torch.linalg.lu(t)
1.21 ms ± 2.23 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
>>> %timeit torch.linalg.lu(t, pivot=False)
488 µs ± 800 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

The performance gain is even larger there. So we have a default algorithm that's slower but stable, and an opt-in to a significantly faster algorithm.

**Comment by rgommers** [comment]
@lezcano @nikitaved it'd be nice to get your opinion on this one. I can't find a PyTorch feature request for LU decomposition with `pivot=False` on CPU, despite it being much faster potentially. Is there a reason why it's only implemented on GPU?

**Comment by lezcano** [comment]
`pivot=False` is often quite risky, as the class of matrices for which the decomposition has a not-so-nice structure (all its principal minors must be away from zero). For example, it does not exist for a matrix `A` with `A[0,0] = 0`. With this limitation, there are very few families of matrices for which this property holds. One of the few families for which this condition holds is positive definite matrices (by Sylvester's criterion), but then the LU decomposition is just good ol' Cholesky, for which we have a specific algorithm that's x2 faster.

Partial pivoting is often a good compromise between stability and speed when compared to, say, full pivoting or other more general decompositions. For this reasons, I don't think that there are many practical cases where you would want to have a regular LU decomposition.

**Comment by nikitaved** [comment]
> For example, it does not exist for a matrix `A` with `A[0,0] = 0`. With this limitation, there are very few families of matrices for which this property holds.
> 
I would not say few, there are plenty of matrices that satisfy such conditions, like full-rank matrices with non-zero diagonals. And in most cases diagonals are pre-conditioned anyway. Besides, `pivot=True` is the default, so I would assume the user is aware of the structure of matrices at hand when `pivot=False` is chosen.

> Partial pivoting is often a good compromise between stability and speed when compared to, say, full pivoting or other more general decompositions. For this reasons, I don't think that there are many practical cases where you would want to have a regular LU decomposition.
There are cases when a single decomposition of a non-symmetric matrix leads to solving linear systems when the rhs comes in a stream. I would suspect that having `pivots` in this case is also detrimental to performance because of the need to re-order rhs.

**Comment by lezcano** [comment]
The diagonal case you mentioned I don't think it's correct, but it's true that there is at least one family of matrices that I know of for which LU decompositions exist, which is column diagonally-dominant matrices.

At any rate, in most cases you do want partial pivoting due to stability reasons, so yeah. This could be implemented in PyTorch, so it shouldn't stop the array API from standarising it. I just think it shouldn't be the default.

Also, in terms of standarisation, I prefer the API we have in PyTorch for `lu_solve` over the one exposed in SciPy. In particular, we don't expose the `trans` parameter, but simply act on the strides of the input tensor to choose the best value for this parameter. We also act on the conj-bit that PyTorch implements to solve `A^H X = B`, but I don't know if NumPy implements that.

**Comment by nikitaved** [comment]
> The diagonal case you mentioned I don't think it's correct, but it's true that there is at least one family of matrices that I know of for which LU decompositions exist, which is column diagonally-dominant matrices.

Agree. Fixed and changed to diagonal dominance which I meant in the first place :)

**Comment by ogrisel** [comment]
>> But it's not clear if pytorch' pivot=False is equivalent to scipy permute_l=True or doing something different.

> It looks like it's doing the exact same thing.

I am not sure but I cannot quickly check since I do not have access to a cuda GPU at the moment:

By reading the docstring of [torch.linalg.lu](https://pytorch.org/docs/stable/generated/torch.linalg.lu.html), I am under the impression that the returned `L` is always lower triangular with ones on the diagonal, whether or not `pivot=False` while this is not the case for `scipy.linalg.lu`: `permute_l=True` would return the product `P @ L` which is unlikely to be lower triangular.

**Comment by ogrisel** [comment]
Also in the case of [randomized_range_finder](https://github.com/scikit-learn/scikit-learn/blob/86541f2b3bc8a96264e265cb810cf79858544340/sklearn/utils/extmath.py#L205) used internally by scikit-learn's `PCA(solver="randomized")`, I confirm that we are interested in applying LU with tall matrices for which the permutation matrix `P` would not fit in memory, hence implicitly precomputing the `P @ L` product is necessary (otherwise one would get a crash instead of a slowdown).

**Comment by lezcano** [comment]
If you have a tall matrix, what you often want to do is to compute the PLU of the transposed matrix and use it to solve the given system.

**Comment by lezcano** [comment]
Also, thank you for digging this up https://github.com/data-apis/array-api/issues/627#issuecomment-1556881760. It looks like

**Comment by oleksandr-pavlyk** [comment]
@lezcano I agree. I think the matrix `P` should not be returned as a matrix, but rather as a vector encoding the permutation. I would suggest that vector to be the image of `range(min(M, N))` under the permutation. That way we could have `(L*U)[p[i], j] == X[i,j]`. The empty `P` array would be understood as the identity permutation.

**Comment by rgommers** [comment]
Cc @ilayn who is currently making changes to `scipy.linalg.lu`

**Comment by lezcano** [comment]
@oleksandr-pavlyk  I was referring to the output of `linalg.lu_factor` which returns the permutation encoded as a vector, but not with the usual encoding. `linalg.lu_factor` is just an interface into BLAS' `getrf`, so it returns the pivots with the same format as BLAS does.

For these functions, a minimal set of functions from which all the others can be implemented is given by:
- `linalg.lu_factor` which would just be modelled after [`scipy.linalg.lu_factor`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.lu_factor.html) (which is BLAS' `getrf`)
- Something along the lines as `linalg.qr_multiply` that could be called `linalg.lu_multiply` (name TBD), with signature `lu_multiply(pivots: Array, B: Array, trans: bool = False)` that, given the pivots in BLAS'  format and a matrix `B`, it computes `P @ B` if `not trans` else `P^T @ B`

With those two, you can implement `lu_solve` in terms of them and `solve_triangular`. You can also implement `linalg.lu` in terms of these + `tril/triu`. IMO, I would be +1 to providing `linalg.lu_solve` even if it's redundant, and I'd suggest not to include `linalg.lu`.

**Comment by ogrisel** [comment]
> It looks to me that we don't want to include the permute_l function, as it's just not useful. The matrix P@L does not have any structure, and it may not be used to solve a system in any way.

In the use case in scikit-learn (https://github.com/data-apis/array-api/issues/627#issuecomment-1556898302) we use `P @ L` as a cheap normalization step (as alternative to QR) in randomized SVD solver after each power iteration. This happens at step 2 of the algorithm presented on page 9 of https://arxiv.org/abs/0909.4061.

While the original paper does not use this LU-based normalization trick, scikit-learn took inspiration from @tygert and @ajtulloch's alternative implementation published in https://github.com/facebookarchive/fbpca. See section 5. of:

http://tygert.com/implement.pdf

We could subsequently confirm with our own [benchmarks](https://github.com/scikit-learn/scikit-learn/blob/main/benchmarks/bench_plot_randomized_svd.py) that the LU-based renormalization would improve the speed-accuracy trade-off of our randomized SVD solver.

I don't see how we could implement this pattern with the `lu_factor` / `lu_solve` API.

**Comment by ilayn** [comment]
The SciPy signature is going to have another keyword here denoted by `p_indices` for lack of a better word to return the permutation as a 1D array. See https://github.com/scipy/scipy/pull/18358

`lu(a, permute_l=False, overwrite_a=False, check_finite=True, p_indices=False)`

and now supports ndarrays of arbitrary size (pertaining to memory constraints). No Fortran code is going to be used and existing infra is being deprecated starting 1.12. 

`overwrite_a` and `check_finite` stuff are LAPACK-isms that does not need any standardizations and at some point I'll propose a deprecation in the future by moving things to Cython native code.

**Comment by lezcano** [comment]
> I don't see how we could implement this pattern with the lu_factor / lu_solve API.

In the API proposed in https://github.com/data-apis/array-api/issues/627#issuecomment-1557278813, there is a specific function to compute `P @ B` for an arbitrary `B` matrix, so that's how you would compute `P @ L`.

On a different note, that's a rather interesting trick for the implementation of that algorithm, which sacrifices a bit of stability, but not too much :) Now, I do not know a priory why would you need `P @ L` rather than just `L`, but that's a story for another day.

**Comment by lezcano** [comment]
>Omitting all pivoting with a rank-deficient matrix for which the randomization may be less than perfect sounds potentially numerically unstable to me.

I think a confusion here is coming from the fact that using `permute_l=False` in scipy seems to hint that it'll perform an LU decomposition without pivoting. This is not the case. As discussed above, there is no way to perform an LU decomposition without pivoting in scipy (and I argue in https://github.com/data-apis/array-api/issues/627#issuecomment-1554786038 that neither should be in this API, at least for now).

With that out of the way, my point is that, when doing an LU decomposition with partial pivoting, `L` and `P @ L` will have exactly the same rank (they will both be full rank, regardless of the rank of the initial matrix). Given this, I argue that using `L` is, a priori, as good a choice as using `P@L`. This may be done by
```python 
X = linalg.lu_factor(A)[0]
L = np.tril(X, k=-1)
k = min(L.shape)
L[:k,:k] += numpy.eye(k)
```
If you want to keep the information of the rank of the input matrix, you should keep the `U`, potentially dividing by the non-zero elements of the diagonal. Or even better, you should look into performing an `LDU` factorisation. 

But again, all this part is deviating from the main topic of conversation. The point here is that all these things (`permute_l=False/True`) with the API proposed in https://github.com/data-apis/array-api/issues/627#issuecomment-1557278813

**Comment by ogrisel** [comment]
> Something along the lines as linalg.qr_multiply that could be called linalg.lu_multiply (name TBD), with signature lu_multiply(pivots: Array, B: Array, trans: bool = False) that, given the pivots in BLAS' format and a matrix B, it computes P @ B if not trans else P^T @ B

One way to cleanly implement this would be to standardize a Linear Operator API as part of the Array API spec but I guess this is out-of-scope for now (since it is kind of related to sparse arrays).

For reference, here are some implementations of linear operators in the ecosystem:

- [scipy.sparse.linalg.LinearOperator](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.LinearOperator.html)
- [tf.linalg.LinearOperator](https://www.tensorflow.org/api_docs/python/tf/linalg/LinearOperator)
- [linear-operator for PyTorch](https://linear-operator.readthedocs.io/en/latest/)

**Comment by ogrisel** [comment]
I experimented in a branch of scikit-learn to use the new scipy API from the `lu_no_fortran` branch to return the permutation data as integer indices (as an alternative to precomputing and returning the `P @ L` product) for the use case of LU-based normalization in randomized SVD following @tygert's paper cited above.

- https://github.com/ogrisel/scikit-learn/pull/16

In conclusion:

- discarding the permutation info is empirically detrimental as @tygert anticipated above,
- using `matrix_p=False` (to return an array of indices instead of `P`) followed by fancy indexing `L` to apply the permutation is fast enough for our use case in scikit-learn (compared to letting scipy precompute the `P @ L` product internally using `permute_l=True`).
- I also checked that the new Cython + LAPACK code in SciPy for the `permute_l=True` case was as fast as the old fortran code.

In my opinion, the standard `xp.linalg.lu` should:

- always return the permutation info as an array of indices for the sake of memory usage,
- not expose the `permute_l=True` option of scipy because it can be quickly reproduced via fancy indexing,
- not expose the `pivot=False` option of PyTorch because it can fail in a data dependent way and not all libraries will be willing to implement this option,
- always return 3 arrays (non polymorphing API).

Here is a summary of the typical snippet following the API I would expect in the standard (if we ignore any API backward compatibility concerns):

```python
X = xp.asarray(np.random.randn(42, 43))
P_indices, L, U = xp.linalg.lu(X)
assert xp.linalg.norm(X - L[P_indices] @ U) < 1e-12
```

This means that scipy would implement this with its new options by default. But maybe `numpy.array_api` could.

Not sure how to deal with backward compat for pytorch either.

**Comment by lezcano** [comment]
It is true that the `lu_multiply` proposed in https://github.com/data-apis/array-api/issues/627#issuecomment-1557278813 can be implemented in terms of advanced indexing (or more generally, a `torch.scatter` in the batched case). Now, what I don't like of the API proposed in https://github.com/data-apis/array-api/issues/627#issuecomment-1559723949 is that it doesn't allow for the most common use case of this API which is calling a `lu_factor` and a `lu_solve` to solve a system of linear equations (or more than one when they share their lhs). This is rather typical when solving matrix ODEs numerically, for example.

I would propose then to just have one function:
`lu_factor(X: Array, p_indices: bool = False) -> Tuple[Array, Array]`
The semantics of this function would be:
- When called with `p_indices=False`, it would return the matrix and the pivots returned by BLAS
- When called with `p_indices=True`, the matrix would still be that returned by BLAS, but the pivots would define a 0-indexed permutation, so that they can be used as indices within an advanced indexing operation.

Note that `p_indices` is what https://github.com/scipy/scipy/pull/18358/ calls `matrix_p`. Choose whichever name you like best.

Uses of this API:
- The most obvious one: Its output can be fed into `lu_solve` as-is to implement a `linalg.solve`
- One may extract the `L` returned by `scipy.linalg.lu` doing as discussed in https://github.com/data-apis/array-api/issues/627#issuecomment-1558897210. The `U` may be got by simply calling `triu` on the output of this function.
- If you set `p_indices=True`, you may compute `P @ B` for an arbitrary matrix `B` by doing `B[perm]` where `perm` is the second array (vector in the non-batched case) that this function returns
- As corollary of the previous point, one may compute the matrix `P` explicitly by doing `P = xp.eye(n)[perm]`
- If one wants to compute `P.T @ B` for a given matrix (as you need to do to compute the adjoint of `lu_solve`) one can do that noting that the inverse permutation of a permutation vector `perm` is given by `argsort(perm)`

**Comment by ilayn** [comment]
Since you folks are working with insane amount of data and tools I can't even begin to understand to handle that data, just a few remarks about the speed of LU or any other linalg stuff in SciPy

 - Quite some performance is lost in `check_finite` across the board. If you have alternative ways to guarantee finiteness then you'll get much more out of that implementation. 
 - Another performance eater is the Fortran-nannying; swapping the C-memory layout to Fortran-layout and back. I tried my best to keep the number of copy operations to 2 (one to lapack one from lapack) but I can't claim expertise. Hence probably I still left quite a bit of performance on the table there. We are working on getting rid of this Fortran ping-pong.
 - There is nothing magical happening in `getrf` and I would cautiously encourage for alternative `getrf` implementations for distributed cases. I will try to make that happen somewhere this summer to reclaim that performance with a pure C-layout implementation.
- The permutation is not necessarily needed to be pushed into the array since it shuffles the memory order and costs performance especially if a copy is created. Instead a permuted solver can go a long way. LAPACK doesn't offer this but as per previous point, rows are contiguous on C-layout and offers quite a nice cache-friendly operations in LU case. 

That is to say, instead of 

```python
Bp = B[p, :]

for r in range(n):
    foo = Bp[r, :] @ bla
``` 

you can do 

```
for r in range(n):
    foo = B[p[r], :] @ bla
```

without any data motion on `B`.

**Comment by ilayn** [comment]
> Note that p_indices is what https://github.com/scipy/scipy/pull/18358 calls matrix_p. Choose whichever name you like best.

While I have the audience, if you have a better naming please do so, we considered 

```
p_indices
perm_as_1d
p_as_1d
p_as_vec
perm_vector
return_1d
return_vector
```

and many other ugly candidates. We went with `matrix_p` out of desperation and to rhyme with `permute_l`

**Comment by ogrisel** [comment]
I think I find `p_indices` more explicit. Or maybe `p_as_indices` to avoid confusion with a local variable that could be used to receive the returned integer array.

**Comment by ogrisel** [comment]
@lezcano I am not sure I understand your proposal in: https://github.com/data-apis/array-api/issues/627#issuecomment-1559831514, in particular "pivots returned by BLAS". Could you please illustrate how your API proposal would feel from an end-user perspective?

**Comment by lezcano** [comment]
My proposal accounts for having 
```python
# p_indices=False
assert xp.linalg.lu_factor(X) == scipy.linalg.lu_factor(X)

# p_indices=True
A, perm = xp.linalg.lu_factor(X, p_indices=True)
assert A == scipy.linalg.lu_factor(X)[0]
assertperm = scipy.linalg.lu(X, matrix_p=True)[2]
```

**Comment by lezcano** [comment]
The full snipped would look like:
```python
# p_indices=False: Solve a linear system
A, P = xp.linalg.lu_factor(X)
sol = xp.linalg_solve(A, P, Y)
assert xp.linalg.norm(X @ sol - Y) < 1e-12

# Reconstruct the full PLU decomposition
A, p_indices = xp.linalg.lu_factor(X, p_indices=True)
L, U = xp.fill_diagonal(xp.tril(A), 1), xp.triu(A)
assert xp.linalg.norm(X - L[p_indices] @ U) < 1e-12
```

The idea behind this API is based in three principles but here is where I may be wrong and it may be helpful for other members of the standarisation committee to chip in:
- In almost all cases, this function will be used through the `p_indices=False` API to solve linear systems or compute a determinant.
- The API should be easily implementable for all libraries (CPU / CUDA) with the current tools (most libraries rely on Blas / cuBLAS / Magma implementations)
- The API should be expressive enough yet compact.

As discussed above, the proposed `xp.linalg.lu_factor` allows for implementing both `scipy.linalg.lu_factor` and `scipy.linalg.lu` using just one function.

I think the conciseness point here is important, as in PyTorch we used to have `lu_factor`, and then [`lu_unpack`](https://pytorch.org/docs/stable/generated/torch.lu_unpack.html). This second one would unpack the `L, U` and the pivots. In my opinion, integrating the unpacking of the returned permutation into a set of indices (i.e. a permutation as defined in mathematics) provides for an even more concise API. 

>find the original suggestion in https://github.com/data-apis/array-api/issues/627#issuecomment-1559723949 way more user-friendly.

The suggestion in that comment does not account for the most common case, which is using the LU decomposition to solve a linear system.

When it comes to the user-friendliness, I believe that the array API is not in the business of offering plenty of auxiliary functions, but to have enough expressibility so that common constructs can be expressed efficiently in it. Auxiliary functions could potentially go into a library that builds on top of the array API. For example, using this explicit construction in your experiments in https://github.com/ogrisel/scikit-learn/pull/16 would save you from instantiating `U` explicitly, halving the peak memory and saving one full read and write over the data, and potentially allowing you to perform the computation of `L`  in place, saving even more memory and one further allocation.

When it comes to efficiency, note that using the NumPy API (that is, without fusion of operations) it's not possible to do better efficency-wise than the described `tril + diagonal_fill + triu` when it comes to computing the `L` and the `U` explicitly.

>Note, that if there is a concern with backward compat with existing libraries and default parameters,

The proposed API is BC with other libraries, as its defaults are the same as those from SciPy and PyTorch

>Also: what is the stance of the Array API standardization committee on "soft" polymorphic functions where the number of array element in a tuple of results does not dependent on the value of the function argument but the dtype and ndim of such arrays do (as would be the case with the p_indices kwargs)

`p_indices` does not change the shape of the output. It simply performs an iterative algorithm on the indices. Here's the algorithm as implemented in PyTorch:
https://github.com/pytorch/pytorch/blob/4882cd08013733a5dbe299871ad7e974bce074b3/aten/src/ATen/native/BatchLinearAlgebraKernel.cpp#L1119-L1128

**Comment by ogrisel** [comment]
Thanks for the explicit summary.

> p_indices does not change the shape of the output. 

I am not sure what you mean by shape. I would expect the following:

```python
>>> A, p_indices = xp.linalg.lu_factor(X, p_indices=True)
>>> p_indices.dtype.kind
'i'
>>> p_indices.ndim
1
```

while:

```python
>>> A, P = xp.linalg.lu_factor(X, p_indices=False)
>>> P.dtype.kind
'f'
>>> P.ndim
2
```

this is what I meant by "soft polymorphic outputs".

**Comment by lezcano** [comment]
When `p_indices=False`, `xp.linalg.lu_factor` would return the same thing as [`scipy.linalg.lu_factor`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.lu_factor.html). In particular, it returns a 1D indexed array that can be fed into `lu_solve`.

If you want to recover the full 2D matrix `P`,  you would do as discussed in the penultimate point in https://github.com/data-apis/array-api/issues/627#issuecomment-1559831514
>[...] one may compute the matrix `P` explicitly by doing `P = xp.eye(n)[perm]`

where `perm = xp.linalg.lu_factor(X, p_indices=True)[1]`

**Comment by ogrisel** [comment]
I new realize that I misunderstood how the `piv` arrays returned by LAPACK actually work. The example in the docstring of `scipy.linalg.lu_factor` seems to be broken.

I find that exposing those low level 1-based indexed pivot info quite confusing anyway. I have the feeling this should not be part of the standard API or at least not with the default options.

**Comment by ilayn** [comment]
I can fix the docs if something is broken. What kind of issue do you have with it? 

And about exposing it, yes that's what I meant in the last part of https://github.com/data-apis/array-api/issues/627#issuecomment-1559360092 about the "mistake". Note that they are pivot "swaps" not permutations.

**Comment by lezcano** [comment]
Which API would you guys then propose that can be implemented with the current tools (BLAS/cuBLAS/cuSolver/MAGMA) that allows for solving linear systems fast, possibly with a common lhs?

**Comment by ilayn** [comment]
My vote goes to `Factorize` class like Julia or Rust implements. The name can be something else but polyalgorithm is a must in this day and age for these operations. `lu_factor`, `chol_solve` etc in my opinion belong to the fortran age. 

If you have an array `A` and pass it to `Factorize` or whatever the name, then you store your array factorization depending on the properties of `A`. It can be discovered at load time, or can be enforced as `Factorize(A)` or `Factorize(A, assume_a='sym')`. Then you can call solve on it and it will use whatever factorization is stored internally. You can repeat solve with different RHS or can perform different ops like `.update()`, `.downdate()`, `det`, `rcond()` in a lazy loading way. It can also hold multiple factorizations. 

If you need explicit factorization then you can still call `lu`, `qr` and so on. That's also sparse and dense agnostic and pretty much what matlab offers as the most famous convenience. You don't focus on the operation itself you focus on the data you provide and the algorithms handle the complication. 

Otherwise we will repeat the same fortran api yet another two decades. That's pretty much what I wanted to do in https://github.com/scipy/scipy/pull/12824 but then I lost my way due to other things. It tries to address the discovery part for solve. The idea is that the linalg APIs have moved on and it is quite inconvenient to fight with lapack api just to get a single operation with internal arrays, convert things from internal lapack representation to usable Scientific Python ways etc. 

We are now trying to reduce our Fortran77 codebase in SciPy which is taking quite some time and after that I am hoping to get back to this again. Anyways, that's a rough sketch but that's kind of my wish to have in the end.

**Comment by ogrisel** [comment]
> I can fix the docs if something is broken. What kind of issue do you have with it?

I thought the examples was broken (I thought I had copied and pasted the snippets and it would fail) but it's actually not the case. Still I think it can be helpful to be more explicit: https://github.com/scipy/scipy/pull/18594

For some reason I thought that I read somewhere that `piv` would be 1-based indexed but it's actually 0-based indexed in scipy. So everything is fine.

**Comment by ogrisel** [comment]
I find the proposal in https://github.com/data-apis/array-api/issues/627#issuecomment-1568892925 interesting. Would it be possible to implement it on top of the existing building blocks available in numpy/scipy, torch and co?

If it can be done, (e.g. maybe in a proof of concept repo), maybe it can help move the spec discussion forward if it is backed by working code.
